src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] arrayList;
+    private int size;
+    private int capacity;
+
     @Override
-    public void add(T value) {
+    protected Object clone() throws CloneNotSupportedException {
+        return super.clone();
+    }
 
+    public void growCapacity() {
+        if (isEmpty()) {
+            arrayList = (T[]) new Object[DEFAULT_CAPACITY];
+            capacity = DEFAULT_CAPACITY;
+        }
+        T[] initialArray = arrayList;
+        arrayList = (T[]) new Object[capacity + capacity / 2];
+        System.arraycopy(initialArray, 0, arrayList, 0, size);
+        capacity = arrayList.length;
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        if (size == capacity) {
+            growCapacity();
+        }
+        arrayList[size++] = value;
+    }
 
+    @Override
+    public void add(T value, int index) {
+        if (index < 0 || index > size) {
```

Comment:
This method could be simplified 

```
checkIndex(index);
resize();
System.arraycopy(elements, index, elements,index + 1, size - index);
elements[index] = value;
size++;
```

Reply:
The method 'checkIndex' has a different implementation.

Reply:
I know, I gave you an example of code. You always can create private method to check if index fits